beacon-chain/slasher/chunks.go

```diff
@@ -0,0 +1,553 @@
+	maxTarget, err := chunkDataAtEpoch(m.params, m.data, validatorIdx, sourceEpoch)
+	if err != nil {
+		return nil, errors.Wrapf(
+			err, "could not get max target for validator %d at epoch %d", validatorIdx, sourceEpoch,
+		)
+	}
+	if targetEpoch < maxTarget {
+		existingAttRecord, err := slasherDB.AttestationRecordForValidator(
+			ctx, validatorIdx, maxTarget,
+		)
+		if err != nil {
+			return nil, errors.Wrapf(
+				err, "could not get existing attestation record at target %d", maxTarget,
+			)
+		}
+		if existingAttRecord != nil {
+			if existingAttRecord.IndexedAttestation.Data.Source.Epoch < sourceEpoch {
+				surroundedVotesTotal.Inc()
+				return &ethpb.AttesterSlashing{
+					Attestation_1: existingAttRecord.IndexedAttestation,
+					Attestation_2: attestation.IndexedAttestation,
+				}, nil
+			}
+		}
+	}
+	return nil, nil
+}
+
+// Update a min span chunk for a validator index starting at the current epoch, e_c, then updating
+// down to e_c - H where H is the historyLength we keep for each span. This historyLength
+// corresponds to the weak subjectivity period of eth2, see:
+// https://github.com/ethereum/eth2.0-specs/blob/dev/specs/phase0/weak-subjectivity.md.
+// This means our updates are done in a sliding window manner. For example, if the current epoch
+// is 20 and the historyLength is 12, then we will update every value for the validator's min span
+// from epoch 20 down to epoch 8 (since 20 - 12 = 8).
+//
+// Recall that for an epoch, e, min((att.target - e) for att in attestations where att.source > e)
+// That is, it is the minimum distance between the specified epoch and all attestation
+// target epochs a validator has created where att.source.epoch > e.
+//
+// Recall that a MinSpanChunksSlice struct represents a single slice for a chunk index
+// from the collection below:
+//
+//                                     val0    val1    val2
+//                                     {  }    {  }    {  }
+//   chunk_0_for_validators_0_to_2 = [[2, 2], [2, 2], [2, 2]]
+//
+//                                     val0    val1    val2
+//                                     {  }    {  }    {  }
+//   chunk_1_for_validators_0_to_2 = [[2, 2], [2, 2], [2, 2]]
+//
+//                            ...
+//
+//                                     val0    val1    val2
+//                                     {  }    {  }    {  }
+//   chunk_N_for_validators_0_to_2 = [[2, 2], [2, 2], [2, 2]]
+//
+// Let's take a look at how this update will look for a real set of min span chunk:
+// For the purposes of a simple example, let's set H = 2, meaning a min span
+// will hold 2 epochs worth of attesting history. Then we set C = 2 meaning we will
+// chunk the min span into arrays each of length 2.
+//
+// So assume we get an epoch 4 and validator 0, then, we need to update every epoch in the span from
+// 4 down to 2. First, we find out which chunk epoch 4 falls into, which is calculated as:
```

Comment:
```suggestion
// 4 down to 3. First, we find out which chunk epoch 4 falls into, which is calculated as:
```
Again, this is because in code we have
```
minEpoch = args.currentEpoch - (m.params.historyLength - 1)
```